libcxx/include/stddef.h

```diff
@@ -7,16 +7,27 @@
 //
 //===----------------------------------------------------------------------===//
 
-#if defined(__need_ptrdiff_t) || defined(__need_size_t) || defined(__need_wchar_t) || defined(__need_NULL) ||          \
-    defined(__need_wint_t)
+#if defined(__need_ptrdiff_t) || defined(__need_size_t) || defined(__need_rsize_t) || defined(__need_wchar_t) ||       \
```

Comment:
I don't think we can unconditionally declare nullptr_t. If e.g. `__need_size_t` is set, then `nullptr_t` shouldn't get declared. I'm not sure if that's for POSIX conformance or what, but it seems to be important when I talk to people implementing low level headers.

Reply:
Actually, I officially don't understand this patch anymore so I would argue we should go for the simpler approach suggested above unless there's a concrete reason why that wouldn't work.

Reply:
The problem is that some headers need to declare _just_ size_t, and it's a problem if _anything_ else gets declared. The way they do that is by setting `__need_size_t` and including stddef.h. I think it's for POSIX reasons but I'm not 100% why it's so important for people to be able to only see pieces of stdarg and stddef.

Reply:
Put another way, the clean version of the header you have above would fail https://github.com/llvm/llvm-project/pull/86252/files#diff-8313c215ed432925f43b313f4a1949bce7ece47e4b23ed1eb31fb5dd8df31a8cR46

Reply:
Yes, my simplified version of the header does fail that test, however I think that test is misguided. Basically, I don't buy this:

> The problem is that some headers need to declare just size_t, and it's a problem if anything else gets declared. 

We would only be defining `nullptr_t` in addition to whatever gets defined in the `#include_next <stddef.h>`. However that's already the status quo and things work, so I don't think this can be a problem.

Reply:
That's not the status quo though. If `__need_size_t` is defined then we don't declare `nullptr_t` right now. I'm very nervous changing that behavior too.

Reply:
How about this -- let's take my patch but not define `nullptr_t` when `__need_size_t` (or one of the others) is defined.

Then I can remove that bit of logic (and define it all the time) in a separate patch. We can test that one on its own.

Reply:
@AaronBallman do you know why the `__need_` macros are so important in the clang stddef.h?

Reply:
Thanks for digging this up. We're checking with some folks internally.

Reply:
The answer we're getting from the Darwin folks is that they can't find a historical reason for it. Basically our current understanding is that on Darwin at least, the most naive implementation (without any `__need_FOO` macros) *should* work. Or at least it would be intended to work.

I don't know about Linuxes though. At the very least, as far as libc++ is concerned, I am quite tempted to close this PR and go with https://github.com/llvm/llvm-project/pull/86843 instead.

Reply:
No one's really sure why it's important for the `__need_` macros to be perfectly strict. Trying out https://github.com/llvm/llvm-project/pull/86843 internally. We're still talking about if we want to keep this PR for the tests and copy stddef.h from that one, or dump the tests and just do 86843 by itself.